### PR TITLE
Convert repl plugin name to single-value string on upgrades

### PR DESCRIPTION
**Pull request by rcritten**
The value was being returned as a list with one value consisting
of bytes. Grab the first element and convert it to a string to
avoid this error on updates:

Add failure missing required attribute "objectclass"

Because the entry wasn't found the update to add nsslapd-pluginbetxn
to the entry tries to add an entirely new entry which consists
only of dn and nsslapd-pluginbetxn, hence the objectclass error.

https://pagure.io/freeipa/issue/8885

Signed-off-by: Rob Crittenden <rcritten@redhat.com>

### CODE REVIEW

**Comment by rcritten** [comment]
Now that I think about it more, I suspect the same function used during install can be used during upgrade since we are only updating, not adding, entries.

**Comment by abbra** [comment]
The refactoring does not work as it introduces a circular dependency:
```
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:513 RUN ['ipa-replica-manage', 'list', '-v', 'replica0.ipa.test']
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:557 Traceback (most recent call last):
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:557   File "/usr/sbin/ipa-replica-manage", line 36, in <module>
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:557     from ipaserver.install import replication, dsinstance, installutils
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:557   File "/usr/lib/python3.9/site-packages/ipaserver/install/replication.py", line 44, in <module>
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:557     from ipaserver.install import installutils
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:557   File "/usr/lib/python3.9/site-packages/ipaserver/install/installutils.py", line 59, in <module>
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:557     from ipaserver.install import certs, service, sysupgrade
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:557   File "/usr/lib/python3.9/site-packages/ipaserver/install/service.py", line 45, in <module>
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:557     from ipaserver.install.ldapupdate import LDAPUpdate
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:557   File "/usr/lib/python3.9/site-packages/ipaserver/install/ldapupdate.py", line 38, in <module>
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:557     from ipaserver.install.replication import get_replication_plugin_name
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:557 ImportError: cannot import name 'get_replication_plugin_name' from partially initialized module 'ipaserver.install.replication' (most likely due to a circular import) (/usr/lib/python3.9/site-packages/ipaserver/install/replication.py)
DEBUG    ipatests.pytest_ipa.integration.host.Host.master.cmd87:transport.py:217 Exit code: 1
```

**Comment by rcritten** [comment]
Thank goodness for CI.

I'll see if I can move the function elsewhere. It is ideal to share it between the two.

File: ipaserver/install/installutils.py [hunk 1]
```diff
@@ -1552,3 +1552,20 @@ def validate_mask():
...
+    # 389-ds cannot handle older alias querying in the plugin
+    # configuration with 'nsslapd-plugin-depends-on-named: ..' attribute
+    #
+    # dirsrv_get_entry: function (dn, attrs) -> str
+    # returns entry dictionary
+    try:
+        entry = dirsrv_get_entry(
+            'cn=Multisupplier Replication Plugin,cn=plugins,cn=config',
+            ['cn'])
+        return str(entry['cn'], encoding='utf-8')
```
Comment (flo-renaud) [inline ipaserver/install/installutils.py:1569]: When called from the installer, entry['cn'] is `b'Multisupplier Replication Plugin'` but when called from the upgrade we get a list of str `['Multisupplier Replication Plugin']`.
The methods used as dirsrv_get_entry are so different that it's difficult to factor the code for both cases.

**Review by flo-renaud — COMMENTED** [review]
Hi Rob
please find inline comments.

**Comment by flo-renaud** [comment]
The issue can be seen with freeipa 4.9.5, I would also add the ipa-4-9 label.

File: ipaserver/install/installutils.py [hunk 2]
```diff
@@ -1552,3 +1552,20 @@ def validate_mask():
...
+    except Exception:
```
[the remark this answers is not in the document]
Comment (rcritten) [reply ipaserver/install/installutils.py:1570]: Yeah, I meant to add that, will do.

File: ipaserver/install/installutils.py [hunk 3]
```diff
@@ -1552,3 +1552,20 @@ def validate_mask():
...
+    # Support renaming of a replication plugin in 389-ds
+    # IPA topology plugin depends on the replication plugin but
+    # 389-ds cannot handle older alias querying in the plugin
+    # configuration with 'nsslapd-plugin-depends-on-named: ..' attribute
+    #
+    # dirsrv_get_entry: function (dn, attrs) -> str
+    # returns entry dictionary
+    try:
+        entry = dirsrv_get_entry(
+            'cn=Multisupplier Replication Plugin,cn=plugins,cn=config',
```
[the remark this answers is not in the document]
Comment (rcritten) [reply ipaserver/install/installutils.py:1567]: Ouch, ok. This is what I get for squashing three use-cases into one. Thanks for catching this.

**Comment by rcritten** [comment]
I need to check my git-fu, I didn't notice this wasn't in ipa-4-9 and thought it was master only. Corrected.

**Comment by rcritten** [comment]
I think I have it worked out now. I tested with both plugin names as the default and both are now working.

I tried to future-proof it when converting the resulting data so will hopefully see it break early if something changes.

**Comment by abbra** [comment]
Please drop po/*.po updates from your PR ;)

**Comment by rcritten** [comment]
Done. I still haven't figured out why sometimes those don't get marked as ignore.

**Comment by flo-renaud** [comment]
Hi @rcritten 
thanks for the updated patch, works for me. I tested the following scenarii:
- [x] install ipa 4.9.4 with ds 2.0.3, upgrade ipa with the patch: keeps MultiMaster name
- [x] install ipa 4.9.4 with ds 2.0.3, upgrade ds to 2.0.5, ipa-server-upgrade: converts to Multisupplier
- [x] install with patch + ds 2.0.3: uses MultiMaster name
- [x] install with patch + ds 2.0.5: uses Multisupplier name
- [x] install a replica with ds 2.0.3: uses Multimaster name
- [x] install a replica with ds 2.05: uses Multisupplier name

**Comment by flo-renaud** [comment]
master:

* 922357b748e8d32f3d87ba7f443a8a4094775b60 Use get_replication_plugin_name in LDAP updater

**Comment by flo-renaud** [comment]
@rcritten the automated backport to ipa-4-9 failed, can you do a manual backport? Thanks.

**Pull request closed without merging**